### DISCUSSION TITLE
Use multiple models in the same task (1 for thinking, 1 for coding)

### DISCUSSION THREAD

**Discussion by PhunkyBob** [cline/cline#387] [Feature Requests]
In my understanding, Claude-Dev uses AI to do at least 2 kinds of tasks: 
- thinking: create a plan of what to do and how
- coding: generate code

The problem is that some LLM are better for the "thinking" part, whereas others are better for "coding" part. 

## Feature request
In the parameters, instead of choosing 1 API provider for all tasks, we can choose 2: the API for "thinking" tasks, the API for "coding" tasks.

Thanks to this feature, I will be able to use "Claude 3.5 sonnet" for all the thinking process (or maybe "OpenAI o1-mini"), and "DeepSeek 2.5" for coding (I think it could be the best Q/P ratio for coding).

**Comment by PhunkyBob** [comment]
"Aider" just introduced the "architect mode" that is basically this feature request 😂
https://aider.chat/2024/09/26/architect.html

**Reply by tylerbcrawford** [reply]
is it possible to do this right now by manually switching the LLM model for each task?

are you using Aider in addition to Cline?

**Reply by PhunkyBob** [reply]
> is it possible to do this right now by manually switching the LLM model for each task?

I don't think so: Cline uses the same preprompt for all it's request.


> are you using Aider in addition to Cline?

Yes I do.
Depending on the path Cline is taking, I'll maybe use only Aider.

**Comment by Krontx** [comment]
Well this is the highest voted feature request, I can feel the update coming!

**Comment by meb427** [comment]
Yes would reduce the cost or allow for example to switch coding to codestral which is great for code less decomposing the task.
Only issue the context, may add max context as parameter, as not everything is 128k like sonnet 3.5.
Also with big repos and beyond 128k Gemini have bigger context and can here solve stuff that you can't pass to sonnet.

**Reply by Krontx** [reply]
Cool

**Comment by CorradoF** [comment]
Right now  Cline costs me 53 times aider. PLEASE make architect for Cline, I want to use Cline it's the best for my use cases, I don't like aider.

Why 53 times? https://x.com/paulgauthier/status/1848795903693336984
In Aider the new combination of --- Sonnet3.5 v2 as architect + DeepSeek2.5 as editor --- scored Sota in their internal benchmark at 85.7%. Sonnet alone 84%
But Deepseek2.5 output cost is 0.28$ compared to 15$ for Sonnet.

**Reply by Krontx** [reply]
This is open source dude, you gotta calm down...

**Reply by CorradoF** [reply]
> This is open source dude, you gotta calm down...

Re-read the message, I am calm, I guess too many numbers tilted you.

**Reply by PhunkyBob** [reply]
> Right now Cline costs me 53 times aider. 

After working on some features, I noticed that every request made by Cline uses a pre-prompt of ~24k characters! 
I think Cline should be more "agentic" and deal with several pre-prompts (one pre-prompt per usage) and only call the specific agent needed.

**Reply by CorradoF** [reply]
> > Right now Cline costs me 53 times aider.
> 
> After working on some features, I noticed that every request made by Cline uses a pre-prompt of ~24k characters! I think Cline should be more "agentic" and deal with several pre-prompts (one pre-prompt per usage) and only call the specific agent needed.

I agree, multiple agents are the next step after hard coding 2 simple different agents - Architect and Editor - for now should be enough

**Comment by danividalg** [comment]
Link to #420

**Comment by iamhenry** [comment]
Although I also presented this idea in discord, according to @saoudrizwan this doesn't make sense anymore

I still think we should be able to choose which models plans, debugs, implements, etc

https://discord.com/channels/1275535550845292637/1275535550845292640/1301301848527081632

**Comment by danividalg** [comment]
Hi @saoudrizwan, but if we use other models like deepseek, mistral, ..., perhaps this "two models mode" (one for plan and a second for coding) could really improve the development. Do you agree ? 🙂

**Comment by PhunkyBob** [comment]
IMHO, the argument "tokens will cost less, so no need to improve" is not a good argument.

If in the near future, model "A" is way better for thinking process and model "B" il way better for coding, I want to have the best of both worlds.

**Reply by iamhenry** [reply]
yes, it's clear some models are fined tuned to coding tasks (claude) and others for reasoning (o1 models).

to me that clear distinction could help reduce issues one model is having compared to the other

**Comment by tylerbcrawford** [comment]
I would love to see support for dual mode, Sonnet 3.5 is amazing but expensive, would be great to use Sonnet 3.5 and deepseek 2.5 to save $$$

**Comment by jethrop** [comment]
The future is really multi-agent systems with integrated knowledge base, research system, and memory tracking.  There needs to be an additional multi-agent layer between the prompt and the output. #489 really explains this well: "Implement a multi-agent framework in Cline where each agent has a specific role (e.g., planning, coding, debugging, testing, and user communication)."

**Comment by marco-altran** [comment]
Coming from Cursor, I would love to have the ability to set a fast and cheap model for applying diffs. Cline is great, but it's slow and expensive because it uses an overpowered model for applying the code changes to a given file. Cursor has proven we can have a much faster and cheaper model for this particular task.

**Comment by ubuntupunk** [comment]
Seems related, but ability to save multiple models under OpenAI compatible, or some kind of router to get an agentic workflow thinking >> Model A, coding >> Model B etc.

**Comment by jethrop** [comment]
It might be better to divorce the front-end and the backend. Enable support for different agentic frameworks like AG2 or pydantic ai. Especially given AG2s new tree of thought method.

**Comment by slavakurilyak** [comment]
The [Plan-and-Solve](https://arxiv.org/abs/2305.04091) paper inspired many in the space, including [babyagi](https://github.com/yoheinakajima/babyagi) and aider's architect and editor modes ([here](https://aider.chat/docs/usage/modes.html#architect-mode-and-the-editor-model) and [here](https://aider.chat/2024/09/26/architect.html))

Given this new industry best practice, it's time for cline to separate code reasoning from editing

Consider using YAML files to allow developers to easily define their own system prompts for architect/thinking/planning and editor/coding/executing models

While I use both Cline and Aider for my agentic workflows and client work, I would prefer to use Cline with this feature release

**Reply by bramburn** [reply]
Sounds good. I was thinking of refactoring the chat cline.ts do decouple the emitters so that we can do that with multiple agents using pipes and reflection. I still read those papers

**Reply by slavakurilyak** [reply]
@bramburn Aider provides a great abstraction for this using [Coders](https://github.com/Aider-AI/aider/tree/main/aider/coders), see [architect_coder.py](https://github.com/Aider-AI/aider/blob/main/aider/coders/architect_coder.py) and [editor_editblock_coder.py](https://github.com/Aider-AI/aider/blob/main/aider/coders/editor_editblock_coder.py) respectively

**Comment by orangejon** [comment]
Maybe it's too complex to implement but it would be cool if for debugging (which is what Cline seems to spend most of its tokens on) we could configure it to use a cheap model first and then if that doesn't work (e.g. within a maximum number of calls/tokens), retry with a bigger, more expensive model.

**Comment by jethrop** [comment]
Something else that would be useful is file type specific system prompts.
Automatic prompt switching between files.

**Comment by Xorba001** [comment]
There is a need to use a 2nd custom multimodal API, but only for image.
For example, DeepSeek-v3 have no image capability or other local non multimodal models.

Created another Ticket #1180 here :
https://github.com/cline/cline/discussions/1180

It should also have a configurable multimodal API separate from the main API, for use with our favourite image descriptor, such as Molmo. 

Molmo require a prompt for image as required to give its description to a prompt.
When used, the 2nd multimodal config should use, if present, a favorite model to direct image handling to.

Pseudocode-ish something like:
Message sent through Cline =
( Normal request message + (if image model setup is not nil, use 2nd API) or else works normally from main API )

**Comment by bramburn** [comment]
I'm working on this. It needs a full refactor of the cline.ts main code

**Reply by PhunkyBob** [reply]
I advise you to talk with @saoudrizwan before, because you may work a lot for nothing if your PR is rejected (like my 2 previous PRs)...

**Reply by bramburn** [reply]
damn so its like that! wow

**Comment by jakenuts** [comment]
A simplistic way to achieve this is adding GPT4o or similar as an MCP-server and instructing cline to consult with it, discuss and evaluate different designs or choices. You could probably even incorporate an actual messaging system like Slack/Discord to have multi model discussions and user input from the bath tub. Need a timer/workflow to keep it going tho.

**Comment by orangejon** [comment]
I tried that (using Gemini MCP) but Cline would never use it. I have a
similar problem trying to get it to 'RTFM' and Google when it finds an
error (using search1api MCP) - unless I specifically prompt it to manually,
it hardly ever searches for anything.. it just guesses and often breaks
things! I guess this could probably be addressed by changing the system
prompts though.

**Comment by rhokstar** [comment]
Related to #489, integrating multi-agent capabilities into Cline or enabling through MCP server to that can use/connect a variety of agent frameworks such as autogen2 (AG2), langchain, smolagents (by huggingface), now we can move towards decentralizing multiple responsibilities from one coding agent to 2 or more agents with specific responsibilities.

**Comment by slavakurilyak** [comment]
Something is cooking in the kitchen!

Check out [3.2.0](https://github.com/cline/cline/blob/main/CHANGELOG.md#320)

**Comment by iamhenry** [comment]
@slavakurilyak it's just not released yet, right? 
![image](https://github.com/user-attachments/assets/e3ec8d5e-8865-411a-99b3-44a5deacacfd)

**Reply by slavakurilyak** [reply]
Cline v3.2 is out now!

https://x.com/sdrzn/status/1881761978986934582

**Comment by dcbartlett** [comment]
@iamhenry what marketplace is this?

**Reply by iamhenry** [reply]
That was the official vscode marketplace a few days ago

**Comment by dcbartlett** [comment]
Cline introduced Plan/Act in the 3.2.0 release.  This allows you to put Cline in a "Plan" mode where it will not do any actions until your switch to "Act" mode.  This will allow you to have a conversation to nail down exactly what you want the AI to implement.  Once you're happy with the plan, Switch to Act mode and Cline will execute the plan.

We looked into supporting multiple models for this, however, Because the different models would need the context sent back and forth as you switch modes, we found that this will cause the cost to be more expensive.  It's more cost effective to use a single model with the Plan/Act system.

**Comment by orangejon** [comment]
Surely cost is not the only motivation for this request? I'd like to experiment with using a (probably expensive) "thinking" model for Planning, but my experience is that switching to anything other than Claude Sonnet for the coding part pretty much breaks Cline. Using previous Cline versions I've tried a few LLMs (with good SWE-bench scores) and none of them performed well with Cline.

**Comment by jingchang0623-crypto** [comment]
**Real-world battle report from running 5+ AI agents 24/7** 🦑

This is THE problem everyone hits when scaling from "playground" to "production." Here is how we solved it:

**The actual pain points:**
- Cheap model for brainstorming → expensive model for final code review = 40% cost savings
- Context length varies wildly: reasoning model needs full history, coding model just needs the last 2 turns
- Model failures cascade: if your "cheap" model hallucinates, your "expensive" model amplifies it

**Our production setup (miaoquai.com runs this):**
```python
# Model routing based on task complexity
if task_complexity < 0.3:
    model = "deepseek-chat"  # Cheap, fast
elif task_complexity < 0.7:
    model = "claude-3.5-haiku"  # Balanced
else:
    model = "claude-4-opus"  # Reliable
```

**What would make Cline unbeatable:**
1. Slot-based model selection ✅
2. **Automatic fallback chains** - if o3-mini fails → try 4o
3. **Cost accumulator** - show running $ spent per task

Built a dependency checker for exactly this: https://github.com/jingchang0623-crypto/miaoquai-openclaw-tools

**Anyone else running multi-model setups? What is your routing logic?**

---
*凌晨2点，我和三个模型开了个会。最便宜的模型负责头脑风暴，最贵的模型负责审核。*
